examples/dcp/stateful_example.py

```diff
@@ -0,0 +1,138 @@
+# based on PyTorch sample https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/examples/stateful_example.py
+
+import os
+
+import torch
+import torch.distributed as dist
+import torch.distributed.checkpoint as dcp
+import torch.multiprocessing as mp
+import torch.nn as nn
+from torch.distributed.checkpoint.state_dict import (
+    _patch_model_state_dict,
+    _patch_optimizer_state_dict,
+)
+from torch.distributed.device_mesh import init_device_mesh
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+
+from s3torchconnector.dcp import S3StorageWriter, S3StorageReader
+
+
+class Model(torch.nn.Module):
+    def __init__(self) -> None:
+        super().__init__()
+        torch.manual_seed(0)
+        self.net1 = nn.Sequential(nn.Linear(8, 16), nn.ReLU())
+        self.net2 = nn.Sequential(nn.Linear(16, 32), nn.ReLU())
+        self.net3 = nn.Linear(32, 64)
+        self.net4 = nn.Sequential(nn.ReLU(), nn.Linear(64, 8))
+
+    def forward(self, x):
+        return self.net4(self.net3(self.net2(self.net1(x))))
+
+    def get_input(self):
+        return torch.rand(8, 8, device="cuda")
+
+
+def _make_stateful(model, optim):
+    _patch_model_state_dict(model)
+    _patch_optimizer_state_dict(model, optimizers=optim)
```

Comment:
I remember in PyTorch's examples a class called `AppState`, which I think does what you're doing here: what about using it instead of those two private functions, so the user could relate better to it?

In fact, why not use https://pytorch.org/tutorials/recipes/distributed_checkpoint_recipe.html#saving instead, for simplicity?

Reply:
I think this example shows more low level approach that is easier to tune up. However, samples that you mentioned easier to understand, I will add them as separate samples.